# Telco Customer Churn Prediction - Data Preparation

## 1. Introduction

This notebook presents the first step in my portfolio project to develop a machine learning model to predict customer churn for a telecommunications company. The data comes from [IBM's sample datasets](https://accelerator.ca.analytics.ibm.com/bi/?perspective=authoring&pathRef=.public_folders%2FIBM%2BAccelerator%2BCatalog%2FContent%2FDAT00148&id=i9710CF25EF75468D95FFFC7D57D45204&objRef=i9710CF25EF75468D95FFFC7D57D45204&action=run&format=HTML&cmPropStr=%7B%22id%22%3A%22i9710CF25EF75468D95FFFC7D57D45204%22%2C%22type%22%3A%22reportView%22%2C%22defaultName%22%3A%22DAT00148%22%2C%22permissions%22%3A%5B%22execute%22%2C%22read%22%2C%22traverse%22%5D%7D), consisting of several excel files containing various attributes.

The goal of this exploration is to understand the data structure, and prepare a consolidated dataset for feature engineering and modeling in subsequent steps of thee workflow.

In [1]:
import numpy as np
import pandas as pd

pd.set_option('display.max_columns', None)

## 2. Loading data and initial examination

In [2]:
# Load datasets into a dictionary for better management
datasets = {
    'demographics': pd.read_excel("../data/raw/Telco_customer_churn_demographics.xlsx"),
    'location': pd.read_excel("../data/raw/Telco_customer_churn_location.xlsx"),
    'population': pd.read_excel("../data/raw/Telco_customer_churn_population.xlsx"),
    'services': pd.read_excel("../data/raw/Telco_customer_churn_services.xlsx"),
    'status': pd.read_excel("../data/raw/Telco_customer_churn_status.xlsx"),
    'churn1': pd.read_excel("../data/raw/CustomerChurn.xlsx"),
    'churn2': pd.read_excel("../data/raw/Telco_customer_churn.xlsx")
}

In [3]:
# Examine each dataset
for name, df in datasets.items():
    print(f"Dataset: {name}")
    print(f"Shape: {df.shape}")
    display(df.head(2))

Dataset: demographics
Shape: (7043, 9)


Customer ID  Count  Gender  Age Under 30 Senior Citizen Married Dependents  \
0  8779-QRDMV      1    Male   78       No            Yes      No         No   
1  7495-OOKFY      1  Female   74       No            Yes     Yes        Yes   

   Number of Dependents  
0                     0  
1                     1

Dataset: location
Shape: (7043, 10)


Location ID Customer ID  Count        Country       State         City  \
0  OXCZEW7397  8779-QRDMV      1  United States  California  Los Angeles   
1  FCCECI8494  7495-OOKFY      1  United States  California  Los Angeles   

   Zip Code                Lat Long   Latitude   Longitude  
0     90022   34.02381, -118.156582  34.023810 -118.156582  
1     90063  34.044271, -118.185237  34.044271 -118.185237

Dataset: population
Shape: (1671, 3)


ID  Zip Code  Population
0   1     90001       54492
1   2     90002       44586

Dataset: services
Shape: (7043, 31)


Service ID Customer ID  Count Quarter Referred a Friend  \
0  IJKDQVSWH3522  8779-QRDMV      1      Q3                No   
1  BFKMZJAIE2285  7495-OOKFY      1      Q3               Yes   

   Number of Referrals  Tenure in Months    Offer Phone Service  \
0                    0                 1      NaN            No   
1                    1                 8  Offer E           Yes   

   Avg Monthly Long Distance Charges Multiple Lines Internet Service  \
0                               0.00             No              Yes   
1                              48.85            Yes              Yes   

  Internet Type  Avg Monthly GB Download Online Security Online Backup  \
0           DSL                        8              No            No   
1   Fiber Optic                       17              No           Yes   

  Device Protection Plan Premium Tech Support Streaming TV Streaming Movies  \
0                    Yes                   No           No              Yes   
1                     No                   No           No               No   

  Streaming Music Unlimited Data        Contract Paperless Billing  \
0              No             No  Month-to-Month               Yes   
1              No            Yes  Month-to-Month               Yes   

    Payment Method  Monthly Charge  Total Charges  Total Refunds  \
0  Bank Withdrawal           39.65          39.65            0.0   
1      Credit Card           80.65         633.30            0.0   

   Total Extra Data Charges  Total Long Distance Charges  Total Revenue  
0                        20                          0.0          59.65  
1                         0                        390.8        1024.10

Dataset: status
Shape: (7043, 12)


Status ID Customer ID  Count Quarter  Satisfaction Score Customer Status  \
0  SWSORB1252  8779-QRDMV      1      Q3                   3         Churned   
1  SNAEQA8572  7495-OOKFY      1      Q3                   3         Churned   

  Churn Label  Churn Value  Churn Score  CLTV Churn Category  \
0         Yes            1           91  5433     Competitor   
1         Yes            1           69  5302     Competitor   

                   Churn Reason  
0  Competitor offered more data  
1  Competitor made better offer

Dataset: churn1
Shape: (7043, 21)


LoyaltyID Customer ID Senior Citizen Partner Dependents  Tenure  \
0     318537  7590-VHVEG             No     Yes         No       1   
1     152148  5575-GNVDE             No      No         No      34   

  Phone Service    Multiple Lines Internet Service Online Security  \
0            No  No phone service              DSL              No   
1           Yes                No              DSL             Yes   

  Online Backup Device Protection Tech Support Streaming TV Streaming Movies  \
0           Yes                No           No           No               No   
1            No               Yes           No           No               No   

         Contract Paperless Billing    Payment Method  Monthly Charges  \
0  Month-to-month               Yes  Electronic check            29.85   
1        One year                No      Mailed check            56.95   

  Total Charges Churn  
0         29.85    No  
1        1889.5    No

Dataset: churn2
Shape: (7043, 33)


CustomerID  Count        Country       State         City  Zip Code  \
0  3668-QPYBK      1  United States  California  Los Angeles     90003   
1  9237-HQITU      1  United States  California  Los Angeles     90005   

                 Lat Long   Latitude   Longitude  Gender Senior Citizen  \
0  33.964131, -118.272783  33.964131 -118.272783    Male             No   
1   34.059281, -118.30742  34.059281 -118.307420  Female             No   

  Partner Dependents  Tenure Months Phone Service Multiple Lines  \
0      No         No              2           Yes             No   
1      No        Yes              2           Yes             No   

  Internet Service Online Security Online Backup Device Protection  \
0              DSL             Yes           Yes                No   
1      Fiber optic              No            No                No   

  Tech Support Streaming TV Streaming Movies        Contract  \
0           No           No               No  Month-to-month   
1           No           No               No  Month-to-month   

  Paperless Billing    Payment Method  Monthly Charges Total Charges  \
0               Yes      Mailed check            53.85        108.15   
1               Yes  Electronic check            70.70        151.65   

  Churn Label  Churn Value  Churn Score  CLTV                  Churn Reason  
0         Yes            1           86  3239  Competitor made better offer  
1         Yes            1           67  2701                         Moved

Some observations:

- `demographics` contains personal observations about customers

- `location` contains geographic information about customers

- `population` gives the population for each zip code (data points not based on customers)

- `services` contains info on the telecom services each customer uses

- `status` contains info about customer account status, cinluding churn

- `churn1` and `churn2` appear to be combined datasets from the ones above with overlapping info

### 2.1 Inspecting primary keys

My strategy will be to combine the first five datasets and check if this contains the information from `churn1` and `churn2`.

Four of these five datasets (excluding `population`) have 7043 rows with a common customer id column. Below I'll check whether these values are unique for each row and the same across datasets, in other words, whether I can use them as primary keys when merging datasets.

In [4]:
primary_datasets = ['demographics', 'location', 'services', 'status']

s = set(datasets[primary_datasets[0]]['Customer ID'])
print(f"Number of unique customer ids in {primary_datasets[0]}: {len(s)}")

for name in primary_datasets[1:]:
    if s == set(datasets[name]['Customer ID']):
        print(f"{name} has the same customer ids")
    else:
        print(f"Warning! {name} does not have the same customer id set")

Number of unique customer ids in demographics: 7043
location has the same customer ids
services has the same customer ids
status has the same customer ids


### 2.2 Standardize column names

Here I'll clean and standardize column names in snake case.

In [5]:
def clean_column_names(df):
    """
    Standardize columns names by removing leading/trailing spaces, converting to lowercase and replacing all non-alphanumeric characters with underscore
    """
    df.columns = (
        df.columns
        .str.strip()
        .str.lower()
        .str.replace(r"[^a-z0-9]+", '_', regex=True)
        .str.replace(r"^_|_$", '', regex=True)
    )
    return df

for name in datasets:
    datasets[name] = clean_column_names(datasets[name])
    
# Verify standardization
for name, df in datasets.items():
    print(f"Columns of {name}: {', '.join(df.columns)}", end='\n\n')

Columns of demographics: customer_id, count, gender, age, under_30, senior_citizen, married, dependents, number_of_dependents

Columns of location: location_id, customer_id, count, country, state, city, zip_code, lat_long, latitude, longitude

Columns of population: id, zip_code, population

Columns of services: service_id, customer_id, count, quarter, referred_a_friend, number_of_referrals, tenure_in_months, offer, phone_service, avg_monthly_long_distance_charges, multiple_lines, internet_service, internet_type, avg_monthly_gb_download, online_security, online_backup, device_protection_plan, premium_tech_support, streaming_tv, streaming_movies, streaming_music, unlimited_data, contract, paperless_billing, payment_method, monthly_charge, total_charges, total_refunds, total_extra_data_charges, total_long_distance_charges, total_revenue

Columns of status: status_id, customer_id, count, quarter, satisfaction_score, customer_status, churn_label, churn_value, churn_score, cltv, churn_categ

## 3. Data quality

Before merging the datasets, I'll check for any obvious issues.

### 3.1 Check for duplicates

In [6]:
for name, df in datasets.items():
    if df.duplicated().sum() == 0:
        print(f"{name}: No duplicated rows")
    else:
        print(f"{name} has {df.duplicated().sum()} rows")

demographics: No duplicated rows
location: No duplicated rows
population: No duplicated rows
services: No duplicated rows
status: No duplicated rows
churn1: No duplicated rows
churn2: No duplicated rows


### 3.2 Check for missing values

In [7]:
def missing_values(name, df):
    """Generate a dataframe containing info on missing values"""
    missing = df.isna().sum()
    missing_pct = round(100 * (missing / len(df)), 2)

    res = pd.DataFrame({
        'dataset': name,
        'column': missing.index,
        'missing_count': missing.values,
        'missing_pct': missing_pct
    })

    return res[res['missing_count'] > 0]

missing_data = pd.concat([missing_values(name, df) for name, df in datasets.items()], ignore_index=True)

missing_data

dataset          column  missing_count  missing_pct
0  services           offer           3877        55.05
1  services   internet_type           1526        21.67
2    status  churn_category           5174        73.46
3    status    churn_reason           5174        73.46
4    churn2    churn_reason           5174        73.46

In [8]:
unique_values = []

for name, col in missing_data[['dataset', 'column']].values:
    unique_values.append((name, col, list(datasets[name][col].unique())))

unique_values_df = pd.DataFrame(unique_values, columns=['dataset', 'column', 'unique_values'])

unique_values_df

dataset          column                                      unique_values
0  services           offer  [nan, Offer E, Offer D, Offer C, Offer B, Offe...
1  services   internet_type                     [DSL, Fiber Optic, Cable, nan]
2    status  churn_category  [Competitor, Dissatisfaction, Price, Other, At...
3    status    churn_reason  [Competitor offered more data, Competitor made...
4    churn2    churn_reason  [Competitor made better offer, Moved, Competit...

Although these categorical columns have a high number of mising values, there isn't a structural issue and they may still contain useful information after transformation. I'll revisit handling missing values during feature engineering in the next notebooks.

## 4. Merging datasets

In this section I'll merge the datasets `demographics`, `location`, `status` and `service` using `customer_id` column as the primary key in each dataset. Afterwards, I'll merge in `population` using the `zip_code` column.

In order to avoid repetitions in columns, I'll check for columns with duplicate names and values during merging.

### 4.1 Merging primary datasets

In [9]:
for name in primary_datasets:
    datasets[name].set_index('customer_id', inplace=True)
    datasets[name].sort_index(inplace=True)

In [10]:
def merge_without_duplicates(df1, df2, df2_name):
    """Merge df1 and df2 while identifying and dropping columns with identical name and values
    """
    common_cols = set(df1.columns) & set(df2.columns)

    # Identify identical columns depends on the index sorting above
    cols_to_drop = [col for col in common_cols if df1[col].equals(df2[col])]

    # Drop identical columns and merge
    df2 = df2.copy().drop(cols_to_drop, axis=1)
    df1 = df1.merge(df2, on='customer_id', how='outer')

    print(f"Merging with {df2_name}")
    print(f"Columns dropped when merging: {cols_to_drop}")
    print(f"Resulting shape: {df1.shape}", end='\n\n')

    return df1

merged_df = datasets[primary_datasets[0]].copy()
print(f"Starting with {primary_datasets[0]}")

for name in primary_datasets[1:]:
    merged_df = merge_without_duplicates(merged_df, datasets[name], name)

Starting with demographics
Merging with location
Columns dropped when merging: ['count']
Resulting shape: (7043, 16)

Merging with services
Columns dropped when merging: ['count']
Resulting shape: (7043, 45)

Merging with status
Columns dropped when merging: ['count', 'quarter']
Resulting shape: (7043, 54)



### 4.2 Integrating `population`

In [11]:
# Reset index to make customer_id a regular column
merged_df = merged_df.reset_index()

merged_df = pd.merge(merged_df, datasets['population'].copy(), on='zip_code', how='left')

# Set customer_id back as index
merged_df.set_index('customer_id', inplace=True)

In [12]:
print(f"Shape: {merged_df.shape}")
display(merged_df.head(2))

Shape: (7043, 56)


count  gender  age under_30 senior_citizen married dependents  \
customer_id                                                                  
0002-ORFBO       1  Female   37       No             No     Yes         No   
0003-MKNFE       1    Male   46       No             No      No         No   

             number_of_dependents location_id        country       state  \
customer_id                                                                
0002-ORFBO                      0  FUGQUJ6597  United States  California   
0003-MKNFE                      0  SIZFEJ5344  United States  California   

                     city  zip_code                lat_long   latitude  \
customer_id                                                              
0002-ORFBO   Frazier Park     93225  34.827662, -118.999073  34.827662   
0003-MKNFE       Glendale     91206  34.162515, -118.203869  34.162515   

              longitude     service_id quarter referred_a_friend  \
customer_id                                                        
0002-ORFBO  -118.999073  MJBAXYDAX5462      Q3               Yes   
0003-MKNFE  -118.203869  NICWXTOGG9486      Q3                No   

             number_of_referrals  tenure_in_months offer phone_service  \
customer_id                                                              
0002-ORFBO                     2                 9   NaN           Yes   
0003-MKNFE                     0                 9   NaN           Yes   

             avg_monthly_long_distance_charges multiple_lines  \
customer_id                                                     
0002-ORFBO                               42.39             No   
0003-MKNFE                               10.69            Yes   

            internet_service internet_type  avg_monthly_gb_download  \
customer_id                                                           
0002-ORFBO               Yes         Cable                       16   
0003-MKNFE               Yes         Cable                       10   

            online_security online_backup device_protection_plan  \
customer_id                                                        
0002-ORFBO               No           Yes                     No   
0003-MKNFE               No            No                     No   

            premium_tech_support streaming_tv streaming_movies  \
customer_id                                                      
0002-ORFBO                   Yes          Yes               No   
0003-MKNFE                    No           No              Yes   

            streaming_music unlimited_data        contract paperless_billing  \
customer_id                                                                    
0002-ORFBO               No            Yes        One Year               Yes   
0003-MKNFE              Yes             No  Month-to-Month                No   

            payment_method  monthly_charge  total_charges  total_refunds  \
customer_id                                                                
0002-ORFBO     Credit Card            65.6          593.3           0.00   
0003-MKNFE     Credit Card            59.9          542.4          38.33   

             total_extra_data_charges  total_long_distance_charges  \
customer_id                                                          
0002-ORFBO                          0                       381.51   
0003-MKNFE                         10                        96.21   

             total_revenue   status_id  satisfaction_score customer_status  \
customer_id                                                                  
0002-ORFBO          974.81  UAAWUJ8685                   3          Stayed   
0003-MKNFE          610.28  URNYXG9268                   5          Stayed   

            churn_label  churn_value  churn_score  cltv churn_category  \
customer_id                                                              
0002-ORFBO           No            0           65  2205            NaN   
00

## 5. Comparing with combined datasets `churn` and `churn2`

Here I'll compare the merged dataset with the two pre-consolidated datasets provided.

In [13]:
merged_df_cols = set(merged_df.columns)
churn1_cols = set(datasets['churn1'].columns)
churn2_cols = set(datasets['churn2'].columns)

print(f"Columns in churn1 but not in merged_df:\n{sorted(churn1_cols - merged_df_cols)}", end='\n\n')
print(f"Columns in churn2 but not in merged_df:\n{sorted(churn1_cols - merged_df_cols)}", end='\n\n')
print(f"Columns in merged_df but not in the other two:\n{sorted(merged_df_cols - (churn1_cols | churn2_cols))}")

Columns in churn1 but not in merged_df:
['churn', 'customer_id', 'device_protection', 'loyaltyid', 'monthly_charges', 'partner', 'tech_support', 'tenure']

Columns in churn2 but not in merged_df:
['churn', 'customer_id', 'device_protection', 'loyaltyid', 'monthly_charges', 'partner', 'tech_support', 'tenure']

Columns in merged_df but not in the other two:
['age', 'avg_monthly_gb_download', 'avg_monthly_long_distance_charges', 'churn_category', 'customer_status', 'device_protection_plan', 'id', 'internet_type', 'location_id', 'married', 'monthly_charge', 'number_of_dependents', 'number_of_referrals', 'offer', 'population', 'premium_tech_support', 'quarter', 'referred_a_friend', 'satisfaction_score', 'service_id', 'status_id', 'streaming_music', 'tenure_in_months', 'total_extra_data_charges', 'total_long_distance_charges', 'total_refunds', 'total_revenue', 'under_30', 'unlimited_data']


The columns of `merged_df` cover the data provided by the other pre-consolidated datasets except for `loyaltyid`. Since this identifier value does not provide any information on the customers, I'll not include it in the merged dataset.

## 6. Exporting dataset & concluding remarks

In this notebook, I have:

- loaded excel files from [IBM's sample datasets](https://accelerator.ca.analytics.ibm.com/bi/?perspective=authoring&pathRef=.public_folders%2FIBM%2BAccelerator%2BCatalog%2FContent%2FDAT00148&id=i9710CF25EF75468D95FFFC7D57D45204&objRef=i9710CF25EF75468D95FFFC7D57D45204&action=run&format=HTML&cmPropStr=%7B%22id%22%3A%22i9710CF25EF75468D95FFFC7D57D45204%22%2C%22type%22%3A%22reportView%22%2C%22defaultName%22%3A%22DAT00148%22%2C%22permissions%22%3A%5B%22execute%22%2C%22read%22%2C%22traverse%22%5D%7D), identified primary datasets and consolidated ones

- identified customer id as the primary key linking datasets

- standardized column names

- checked for structural issue such as duplicate rows and missing values

- merged primary datasets and ensured all data is contained in the merged dataset

In [14]:
print(f"Shape: {merged_df.shape}")
merged_df.head()

Shape: (7043, 56)


count  gender  age under_30 senior_citizen married dependents  \
customer_id                                                                  
0002-ORFBO       1  Female   37       No             No     Yes         No   
0003-MKNFE       1    Male   46       No             No      No         No   
0004-TLHLJ       1    Male   50       No             No      No         No   
0011-IGKFF       1    Male   78       No            Yes     Yes         No   
0013-EXCHZ       1  Female   75       No            Yes     Yes         No   

             number_of_dependents location_id        country       state  \
customer_id                                                                
0002-ORFBO                      0  FUGQUJ6597  United States  California   
0003-MKNFE                      0  SIZFEJ5344  United States  California   
0004-TLHLJ                      0  RZDAXJ8786  United States  California   
0011-IGKFF                      0  MGKGVM9555  United States  California   
0013-EXCHZ                      0  FJLSME1564  United States  California   

                     city  zip_code                lat_long   latitude  \
customer_id                                                              
0002-ORFBO   Frazier Park     93225  34.827662, -118.999073  34.827662   
0003-MKNFE       Glendale     91206  34.162515, -118.203869  34.162515   
0004-TLHLJ     Costa Mesa     92627  33.645672, -117.922613  33.645672   
0011-IGKFF       Martinez     94553  38.014457, -122.115432  38.014457   
0013-EXCHZ      Camarillo     93010  34.227846, -119.079903  34.227846   

              longitude     service_id quarter referred_a_friend  \
customer_id                                                        
0002-ORFBO  -118.999073  MJBAXYDAX5462      Q3               Yes   
0003-MKNFE  -118.203869  NICWXTOGG9486      Q3                No   
0004-TLHLJ  -117.922613  DCSKWRXAI3251      Q3                No   
0011-IGKFF  -122.115432  ZEOATALAE9483      Q3               Yes   
0013-EXCHZ  -119.079903  MVMZRJAHU9423      Q3               Yes   

             number_of_referrals  tenure_in_months    offer phone_service  \
customer_id                                                                 
0002-ORFBO                     2                 9      NaN           Yes   
0003-MKNFE                     0                 9      NaN           Yes   
0004-TLHLJ                     0                 4  Offer E           Yes   
0011-IGKFF                     1                13  Offer D           Yes   
0013-EXCHZ                     3                 3      NaN           Yes   

             avg_monthly_long_distance_charges multiple_lines  \
customer_id                                                     
0002-ORFBO                               42.39             No   
0003-MKNFE                               10.69            Yes   
0004-TLHLJ                               33.65             No   
0011-IGKFF                               27.82             No   
0013-EXCHZ                                7.38             No   

            internet_service internet_type  avg_monthly_gb_download  \
customer_id                                                           
0002-ORFBO               Yes         Cable                       16   
0003-MKNFE               Yes         Cable                       10   
0004-TLHLJ               Yes   Fiber Optic                       30   
0011-IGKFF               Yes   Fiber Optic                        4   
0013-EXCHZ               Yes   Fiber Optic                       11   

            online_security online_backup device_protection_plan  \
customer_id                                                        
0002-ORFBO               No           Yes                     No   
0003-MKNFE               No            No                     No   
0004-TLHLJ               No            No                    Yes   
0011-IGKFF               No           Yes                    Yes   
0013-EXCHZ               No        

In [15]:
merged_df.to_csv("../data/processed/telco_merged.csv")

Remarks:

- The merged dataset has no duplicate rows.

- There are missing values in specific categorical columns `offer`, `internet_type`, `churn_category`, `churn_reason` but they may be informative, for instance, as no internet service. I'll handle missing values in the next notebook.

- Some attributes appear in multiple formats, for instance, `age`, `under_30`, `senior_citizen`. I'm keeping all versions for now, I'll remove redundant ones in the next notebook.

In the next notebook, I'll focus on feature engineering including handling missing values, repeated or redundant features, variable transformation and data leakage as well as creating new features, exploratory data analysis, correlation and outliers.
